openslides/agenda/projector.py

```diff
@@ -1,54 +1,78 @@
-        pk = self.config_entry.get('id')
-        if pk is not None:
-            # Children slide.
-            if not Item.objects.filter(pk=pk).exists():
-                raise ProjectorException('Item does not exist.')
+    # Build a dict from an item_id to all its children
+    children: Dict[int, List[int]] = defaultdict(list)
+    for item in sorted(all_data['agenda/item'].values(), key=lambda item: item['weight']):
+        if item['type'] == 1:  # only normal items
+            children[item['parent_id'] or 0].append(item['id'])
+
+    def get_children(item_ids: List[int]) -> List[Tuple[int, List[Tuple[int, List[Any]]]]]:
+        return [(all_data['agenda/item'][item_id]['title'], get_children(children[item_id])) for item_id in item_ids]
+
+    return get_children(children[parent_id])
 
 
-class ListOfSpeakersSlide(ProjectorElement):
+def items(config: Dict[str, Any], all_data: AllData) -> Dict[str, Any]:
     """
-    Slide definitions for Item model.
-    This is only for list of speakers slide. You have to set 'id'.
+    Item list slide.
+
+    Returns all root items or all children of an item.
     """
-    name = 'agenda/list-of-speakers'
+    root_item_id = config.get('id') or None
+    show_tree = config.get('tree') or False
 
-    def check_data(self):
-        if not Item.objects.filter(pk=self.config_entry.get('id')).exists():
-            raise ProjectorException('Item does not exist.')
+    if show_tree:
+        items = get_tree(all_data, root_item_id or 0)
+    else:
+        items = []
+        for item in sorted(all_data['agenda/item'].values(), key=lambda item: item['weight']):
+            if item['parent_id'] == root_item_id and item['type'] == 1:
+                items.append(item['title'])
 
-    def update_data(self):
-        return {'agenda_item_id': self.config_entry.get('id')}
+    return {'items': items}
 
 
-class CurrentListOfSpeakersSlide(ProjectorElement):
+def list_of_speakers(config: Dict[str, Any], all_data: AllData) -> Dict[str, Any]:
     """
-    Slide for the current list of speakers.
+    List of speakers slide.
+
+    Returns all usernames, that are on the list of speaker of a slide.
     """
-    name = 'agenda/current-list-of-speakers'
+    item_id = config.get('id') or 0  # item_id 0 means current_list_of_speakers
+
+    # TODO: handle item_id == 0
+
+    try:
+        item = all_data['agenda/item'][item_id]
+    except KeyError:
+        return {'error': 'Item {} does not exist'.format(item_id)}
+
+    user_ids = []
+    for speaker in item['speakers']:
+        user_ids.append(speaker['user'])
+    return {'user_ids': user_ids}
 
 
-def get_projector_elements() -> Generator[Type[ProjectorElement], None, None]:
-    yield ItemListSlide
-    yield ListOfSpeakersSlide
-    yield CurrentListOfSpeakersSlide
+def register_projector_elements() -> None:
+    register_projector_element('agenda/item-list', items)
+    register_projector_element('agenda/list-of-speakers', list_of_speakers)
```

Comment:
Please readd the current-list-of-speakers projector element to be more consistent with the old code. We can remove it later if necessary.